python/src/robyn/reporting/onepager_reporting.py

```diff
@@ -162,114 +170,142 @@ def _get_model_info(self, solution_id: str) -> Dict[str, str]:
-            },
-            'fitted_vs_actual': {
-                'position': gs[0, 1],
-                'title': 'Actual vs. Predicted Response',
-                'func': lambda ax: pareto_viz.generate_fitted_vs_actual(solution_id, ax) if pareto_viz else None
-            },
-            'diagnostic': {
-                'position': gs[1, 1],
-                'title': 'Fitted vs. Residual',
-                'func': lambda ax: pareto_viz.generate_diagnostic_plot(solution_id, ax) if pareto_viz else None
-            },
-            'immediate_carryover': {
-                'position': gs[2, 0],
-                'title': 'Immediate vs. Carryover Response Percentage',
-                'func': lambda ax: pareto_viz.generate_immediate_vs_carryover(solution_id, ax) if pareto_viz else None
-            },
-            'adstock': {
-                'position': gs[2, 1],
-                'title': 'Adstock Rate Analysis',
-                'func': lambda ax: pareto_viz.generate_adstock_rate(solution_id, ax) if pareto_viz else None
-            },
-            'bootstrap': {
-                'position': gs[3, 0],
-                'title': 'Bootstrapped Performance Metrics',
-                'func': lambda ax: cluster_viz.generate_bootstrap_confidence(solution_id, ax) if cluster_viz else None
-            },
-            'response_curves': {
-                'position': gs[3, 1],
-                'title': 'Response Curves and Mean Spends by Channel',
-                'func': lambda ax: response_viz.generate_response_curves(solution_id, ax)
+        Args:
+            solution_id: Solution ID to generate plots for
+            plots: List of PlotType enums specifying which plots to generate
+            gs: GridSpec for plot layout
+            
+        Raises:
+            ValueError: If invalid PlotType is provided
+            TypeError: If plots argument contains non-PlotType elements
+        """
+        try:
+            # Validate plot types
+            for plot in plots:
+                if not isinstance(plot, PlotType):
+                    logger.error(f"Invalid plot type provided: {plot}. Must be PlotType enum.")
+                    raise TypeError(f"Plot type must be PlotType enum, got {type(plot)}")
+            
+            # Initialize visualizers
+            pareto_viz = ParetoVisualizer(self.pareto_result, self.adstock, self.mmm_data) if self.adstock else None
+            cluster_viz = ClusterVisualizer(self.pareto_result, self.clustered_result, self.mmm_data) if self.clustered_result else None
+            response_viz = ResponseVisualizer(self.pareto_result, self.mmm_data)
+            transfor_viz = TransformationVisualizer(self.pareto_result, self.mmm_data)
+            
+            # Add space at top for title
+            gs.update(top=0.85)
+            
+            # Define plot positions and functions using PlotType enum directly
+            plot_config = {
+                PlotType.SPEND_EFFECT: {
```

Comment:
Maybe I wasn't clear in my other comment but can we get rid of the hard coded keys from the config here? Lets add this to the backlog for code cleanup:
Essentially create a dataclass (can be left in this file) which looks like this:
@dataclass
class PlotConfig:
plotType: PlotType
position: whatever that datatype is
title: str

REMOVE func from here altogether its just hard to read and follow. So put this logic directly where you're going to want to call the plot functions (maybe in some switch statement at line 256)

Reply:
Can be resolved if you understand my ask here otherwise its good to go from my side. I can stamp it once you mark it resolved.

Reply:
This code needs further refactor. Now that we allow user to select what plots he/she wants in one pager hardcoding the positions will cause empty space between graphs like this.
<img width="897" alt="Screenshot 2024-11-14 at 6 23 46 AM" src="https://github.com/user-attachments/assets/e7075594-8e86-4367-b364-fe5234aee2da">
The plan is to dynamically pic the plot position.(TODO)

Reply:
Made changes to dynamically chose the plot placing in this PR. further refactoring to move out the config and make the code more readble can be done as part of a different PR.